src/compartment.c

```diff
@@ -239,22 +253,86 @@ void
-    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size + new_comp->scratch_mem_stack_size);
+    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size);
     // TODO bounds double-check
     new_comp->ddc = new_ddc;
 }
 
+/* For a given Compartment `new_comp`, an address `intercept_target` pointing
+ * to a function found within the compartment which we would like to intercept,
+ * and a `intercept_data` struct representing information to perform the
+ * intercept, synthesize and inject intructions at the call point of the
+ * `intercept_target`, in order to perform a transition out of the compartment
+ * to call the appropriate function with higher privileges.
+ */
 void
 comp_add_intercept(struct Compartment* new_comp, uintptr_t intercept_target, struct func_intercept intercept_data)
 {
+    // TODO check whether negative values break anything in all these generated functions
     printf("Found `%s` func at %p.\n", intercept_data.func_name, (void*) intercept_target);
-    int offset = (intercept_data.redirect_func - intercept_target) / 4;
-    assert(offset < (1 << 27));
-    offset &= (1 << 26) - 1;
-    const int arm_b_instr_mask = 0b101 << 26;
-    uint32_t instr_binary = arm_b_instr_mask | offset;
-    int* init_addr_ptr = (void*) intercept_target;
-    struct intercept_patch new_patch = { init_addr_ptr, instr_binary };
+    int32_t new_instrs[INTERCEPT_INSTR_COUNT];
+    size_t new_instr_idx = 0;
+    uintptr_t comp_manager_cap_addr = new_comp->manager_caps + new_comp->active_manager_caps_count * sizeof(void* __capability); // TODO
+
+    const int32_t arm_function_target_register = 0b01010; // use `x10` for now
+    const int32_t arm_transition_target_register = 0b01011; // use `x11` for now
+
+    // movz x0, $target_fn_addr:lo16
+    // movk x0, $target_fn_addr:hi16
+    /*assert(intercept_target < (1 << 32));*/
+    const int32_t arm_movz_instr_mask = 0b11010010100 << 21;
+    const int32_t arm_movk_instr_mask = 0b11110010101 << 21;
+    const ptraddr_t target_address_lo16 = (intercept_data.redirect_func & ((1 << 16) - 1)) << 5;
+    const ptraddr_t target_address_hi16 = (intercept_data.redirect_func >> 16) << 5;
+    const int32_t arm_movz_intr = arm_movz_instr_mask | target_address_lo16 | arm_function_target_register;
+    const int32_t arm_movk_intr = arm_movk_instr_mask | target_address_hi16 | arm_function_target_register;
+    new_instrs[new_instr_idx++] = arm_movz_intr;
+    new_instrs[new_instr_idx++] = arm_movk_intr;
+
+    /* `ldpbr` instr generation */
+    // TODO do we have space to insert these instructions?
+    // TODO what if we need to jump more than 4GB away?
+    // Use `adrp` to get address close to address of manager capability required
+    // adrp x11, $OFFSET
+    const int32_t arm_adrp_instr_mask = 0b10010000 << 24;
+    const ptraddr_t target_address = (comp_manager_cap_addr >> 12) - (intercept_target >> 12);
+    /*assert(target_address < (2 << 32));*/
```

Comment:
I've added the assert back in (https://github.com/capablevms/CHERI-ELF-comp/pull/2/commits/bbfc6a9a52d13958bbf0e358807b0e428aa324c1), and changed it to `1 << 32`, as I believe that is the correct value.

Reply:
Another C gotcha: `1 << 32` is (usually) undefined because `1` is an `int`, and the shift amount is too large. You'll need to write `((ptraddr_t)1 << 32)` or similar.